### PR TITLE
feat: honour anonId only from header

### PR DESCRIPTION
**Pull request by cisse21**
# Description

https://www.notion.so/rudderstacks/userID-anonymousID-ec2d067531ac42aba5f84533f32b9135

## Notion Ticket

https://www.notion.so/rudderstacks/8aac9087df644365acdf64e28e290153?v=6e06b0a5ade24f0aa5ffe05dc2972e84&p=97fe15a422c84dd7b906552a54983550&pm=s

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: gateway/gateway.go [hunk 1]
```diff
@@ -515,10 +518,10 @@ func (gateway *HandleT) userWebRequestWorkerProcess(userWebRequestWorker *userWe
 				userIDFromReq := strings.TrimSpace(vjson.Get("userId").String())
 				if builtUserID == "" {
 					if anonIDFromReq != "" {
-						builtUserID = anonIDFromReq + DELIMITER + userIDFromReq
+						builtUserID = userIDHeader + DELIMITER + anonIDFromReq + DELIMITER + userIDFromReq
 					} else {
 						// Proxy gets the userID from body if there is no anonymousId in body/header
-						builtUserID = userIDFromReq + DELIMITER + userIDFromReq
+						builtUserID = userIDHeader + DELIMITER + userIDFromReq + DELIMITER + userIDFromReq
 					}
```
Comment (lvrach) [inline gateway/gateway.go:525]: How much effort would it require to write tests for this section? Maybe isolating this to a function would make it possible or easier.

**Review by fracasula — COMMENTED** [review]
Can we have a test that given an HTTP request with the correct header we check that the job in the DB has `user_id` populated accordingly?

**Review by BonapartePC — APPROVED** [review]

File: gateway/gateway.go (hunk 1 above)
Comment (cisse21) [reply gateway/gateway.go:525]: We don't have such tests for gateway :/ I can isolate this part and add it though if it makes sense

Comment (fracasula) [reply gateway/gateway.go:525]: Can't we copy from `multi_tenant_test.go` and have the app start in GATEWAY mode only, send an event with the correct header and then check the `user_id` on the DB? Or you think it's going to be more complex?

File: gateway/gateway.go [hunk 2]
```diff
@@ -515,10 +518,10 @@ func (gateway *HandleT) userWebRequestWorkerProcess(userWebRequestWorker *userWe
 				userIDFromReq := strings.TrimSpace(vjson.Get("userId").String())
 				if builtUserID == "" {
 					if anonIDFromReq != "" {
-						builtUserID = anonIDFromReq + DELIMITER + userIDFromReq
+						builtUserID = userIDHeader + DELIMITER + anonIDFromReq + DELIMITER + userIDFromReq
```
Comment (atzoum) [inline gateway/gateway.go:521]: I am not sure I understand why we need a composite key containing three different keys, coming both from the request header and the message's content

File: multi_tenant_test.go [hunk 3]
```diff
@@ -314,6 +314,13 @@ func testMultiTenantByAppType(t *testing.T, appType string) {
 			).Scan(&count)
 			require.NoError(t, err)
 			require.Equal(t, 1, count)
+
+			var userId string
```
Comment (atzoum) [inline multi_tenant_test.go:318]: tests should be single purpose too :)
You can imagine what can happen as we are adding more and more scenarios in a single test

Comment (cisse21) [reply multi_tenant_test.go:318]: Yeah I am adding integration tests under the gateway folder. Will clean this up from here

**Pull request merged**